ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/EstablishedPeers.hs

```diff
@@ -74,49 +76,48 @@ belowTargetLocal actions
                    localRootPeers,
                    knownPeers,
                    establishedPeers,
+                   inProgressDemoteWarm,
                    inProgressPromoteCold
                  }
 
-    -- Are we below the target for number of /local/ root peers that are
-    -- established? Our target for established local root peers is all of them!
-    -- However we still don't want to go over the number of established peers
-    -- or we'll end up in a cycle.
-  | numLocalEstablishedPeers + numLocalConnectInProgress
-  < targetNumberOfLocalPeers
-
-    -- Are there any /local/ root peers that are cold we could possibly pick to
-    -- connect to? We can subtract the local established ones because by
-    -- definition they are not cold and our invariant is that they are always
-    -- in the connect set. We can also subtract the in progress ones since they
-    -- are also already in the connect set and we cannot pick them again.
-  , numLocalAvailableToConnect - numLocalEstablishedPeers
-                               - numLocalConnectInProgress > 0
-  --TODO: switch style to checking if the set is empty
+    -- Are there any groups of local peers that are below target?
+  | not (null groupsBelowTarget)
+    -- We need this detailed check because it is not enough to check we are
+    -- below an aggregate target. We can be above target for some groups
+    -- and below for others.
+
+    -- Are there any groups where we can pick members to promote?
+  , let groupsAvailableToPromote =
+          [ (numMembersToPromote, membersAvailableToPromote)
+          | let availableToPromote =
+                  localAvailableToConnect
+                     Set.\\ localEstablishedPeers
+                     Set.\\ localConnectInProgress
+                     Set.\\ localDemoteInProgress
```

Comment:
Should we take into account `localDemoteInProgress`?  This means we can go above the target while we are demoting some peers, which I think we should avoid. Note that below in `numMembersToPromote` we don't take demotions to `Cold` into account.

Reply:
That's right! Thank you for noticing this!